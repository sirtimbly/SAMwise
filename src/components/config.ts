import * as mithril from "mithril";

