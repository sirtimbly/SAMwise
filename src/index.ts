
// var assignmentModel = {
//   fulltime: 40,
//   parttime: 10,
//   total: 50,
//   requested: 80,
//   isEditing: false
// }
// assignmentModel.present = function(data) {
//   console.log("new data presented to model", data)
//   if (data !== undefined) {
//     var data_ = {};
//     data_.fulltime = data.fulltime || assignmentModel.fulltime;
//     data_.parttime = data.parttime || assignmentModel.parttime;
//     data_.requested = data.requested || assignmentModel.requested;
//   } 
//   console.log("data_", data_)
//   var newTotal = Number(data_.fulltime) + Number(data_.parttime);
//   console.log("newtotal", newTotal)
//   if (newTotal <= assignmentModel.requested) {
//       assignmentModel.fulltime = data_.fulltime;
//       assignmentModel.parttime = data_.parttime;
//       assignmentModel.total = newTotal;
//    } else {
     
//    }
  
//   console.log('updated model, calling state.render ', assignmentModel)
//   state.render(assignmentModel);
// }


// var assignmentActions = {}

// assignmentActions.load = function(present) {
//   console.log('loading')
  
//     //pretending we made an api call and it's now complete
//     console.log('calling present')
//     present({
//       fulltime:20,
//       parttime:30,
//       requested:80
//     })
//   }


// assignmentActions.change = function({
//   value,
//   target
// }, present) {
//   console.log("action:change", {value, target})
//   present = present || assignmentModel.present; //this makes event listeners cleaner but binds us
//   var data_ = {};
//   data_[target] = value;
//   present(data_);
// }

// var vm = {};

// vm.reviewData = function(model) {
//   return {
//     fulltime: model.fulltime || 0,
//     parttime: model.parttime || 0,
//     total: Number(model.fulltime) + Number(model.parttime)
//   }
// }

// var Slider = {
//   view: function(vnode) {
//     return m("main.border.rounded.m1.p1", [
//       m("h1", {
//         class: "title"
//       }, vnode.attrs.count + " " + vnode.attrs.name + " shifts assigned."),
//       m("input", {
//         type: "range",
//         min: "0",
//         max: "100",
//         value: vnode.attrs.count,
//         onchange: function(e) {
//           assignmentActions.change({
//             value: e.target.value,
//             target: vnode.attrs.name
//           })
//         }
//       }),
//     ])

//   }
// }

// var Edit = {
//   view: function(vnode) {
//     console.log(vnode.attrs)
//     return m('div.border.p2.m2', [
//       m(Slider, {count: vnode.attrs.fulltime, name: 'fulltime'}),
//       m(Slider, {count: vnode.attrs.parttime, name: 'parttime'})
//     ]);
//   }
// }

// var Review = {
//   view: function(vnode) {
//     return m("table.table", [
//       m("tbody", [
//         m("tr", [
//           m("td", "Full Time Slots"),
//           m("td", vnode.attrs.fulltime)
//         ]),
//         m("tr", [
//           m("td", "Fullt Time Slots"),
//           m("td", vnode.attrs.fulltime)
//         ]),
//         m("tr.text-bold.bg-blue", [
//           m("td", "Total"),
//           m("td", vnode.attrs.total)
//         ])
//       ])
//     ])
//   }
// }

// var state = {};

// state.render = function(model) {
//   console.log("state.render");
//   state.representation(model);
//   state.nextAction(model);
// }

// state.representation = function(model) {
//   //check values to determine current app state
//   console.log('sate.representation -- finally time for the view')
//   editView(model);
// }

// state.nextAction = function(model) {
//   console.log('state.nextaction')
// }

// function editView(model) {
//   console.log("running editview")
//   m.render(document.getElementById('results'), m(Edit, model));
// }

// console.log("window onload");
// state.render(assignmentActions.load(assignmentModel.present))
//   //assignmentModel.present(assignmentActions.load()), state.renderer()
//   //assignmentActions.load(assignmentModel.present(state.renderer))


// //essentially state control
// // m.route(results, "/edit", {
// //   "/edit": m(Edit, assignmentModel),
// //   "/review": m(Review, vm.reviewData(assignmentModel))
// // })
