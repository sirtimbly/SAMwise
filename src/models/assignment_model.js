"use strict";
exports.__esModule = true;
/**
 * AssignmentProps
 */
var AssignmentProps = (function () {
    function AssignmentProps(data) {
        if (data) {
            this.fulltime = data.fulltime || 0;
            this.parttime = data.parttime || 0;
            this.requested = data.requested || 0;
            this.total = data.total || 0;
            this.isEditing = data.isEditing || 0;
        }
    }
    return AssignmentProps;
}());
exports.AssignmentProps = AssignmentProps;
/**
 * AssignmentModel
 */
var AssignmentModel = (function () {
    function AssignmentModel() {
        var _this = this;
        this.present = function (data) {
            var data_ = new AssignmentProps();
            if (data !== undefined) {
                data_.fulltime = data.fulltime || _this.props.fulltime;
                data_.parttime = data.parttime || _this.props.parttime;
                data_.requested = data.requested || _this.props.requested;
                if (data.isEditing !== undefined) {
                    _this.props.isEditing = data.isEditing;
                }
            }
            var newTotal = Number(data_.fulltime) + Number(data_.parttime);
            console.log("newtotal", newTotal);
            if (newTotal <= _this.props.requested) {
                _this.props.fulltime = data_.fulltime;
                _this.props.parttime = data_.parttime;
            }
            else {
                if (data.fulltime && !data.parttime) {
                    _this.props.fulltime = data_.fulltime - (newTotal - _this.props.requested);
                }
                if (data.parttime && !data.fulltime) {
                    _this.props.parttime = data_.parttime - (newTotal - _this.props.requested);
                }
            }
            //always recalculate total
            _this.props.total = _this.props.fulltime + _this.props.parttime;
            //finally we close the functional reactive loop
            _this.stateRenderer(_this.props);
        };
        this.props = new AssignmentProps({
            fulltime: 20,
            parttime: 30,
            requested: 80
        });
    }
    return AssignmentModel;
}());
exports.AssignmentModel = AssignmentModel;
