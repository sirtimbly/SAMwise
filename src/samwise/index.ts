export { IActions, ViewActions } from "./Actions";
export { App } from "./App";
export { AppState, IState } from "./State";
export { Model } from "./Model";
export { ComponentMap } from "./ComponentMap";
export { IView } from "./View";
