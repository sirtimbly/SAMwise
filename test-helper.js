"use strict";
exports.__esModule = true;
var htmltidy = require("tidy-html5");
var m = require("mithril");
exports.defaultHtmlTidyOptions = {
    "show-body-only": true,
    "drop-empty-elements": false,
    "doctype": "omit",
    "indent": true,
    "quiet": true,
    "show-warnings": false,
    // Recognize SVG tags:
    "new-blocklevel-tags": ["svg", "defs"],
    "new-inline-tags": ["path", "polyline", "line", "polygon"]
};
exports.tidy = function (vnodes, htmltidyOptions) {
    if (htmltidyOptions === void 0) { htmltidyOptions = exports.defaultHtmlTidyOptions; }
    var htmlElement = document.createElement("div");
    m.render(htmlElement, vnodes);
    var html = htmlElement.innerHTML;
    return htmltidy.tidy_html5(html, htmltidyOptions);
};
