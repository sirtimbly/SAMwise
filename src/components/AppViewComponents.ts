import * as m from 'mithril';
// import Mithril from 'mithril';
import { App, State, Model, Action } from '../samwise';
import { UITab, UIProgressBar } from './UIAtoms';
import { UISlider } from './UIMolecules';
import { AssignmentModel, AssignmentProps } from '../models/assignment_model';
/* ============================================ */
/**
 * App View Components (these functions make the actual UI layer updates). 
 * They are implementation specific and tightly coupled to the actions above from the app.actions object.
 * + edit
 *  - slider
 * + review
 */
interface VState {} 


export class AppViewComponents {
    app:App;
    constructor (app:App) {
        this.app = app;
    }
    /**
     * Tabs component is coupled to the state model so we define it here instead of in the UIMolecules file
     */
    public Tabs = {
        view: (vnode) => {
            return m("div.flex.flex-row.mx3.mt1", [
                UITab('Widget Production Plan', vnode.attrs.isEditing, this.app.actions.switchToEdit),
                UITab('Review', !vnode.attrs.isEditing, this.app.actions.switchToReview)
            ])
        }
    } as m.Component<AssignmentProps,VState> & VState 

    

    /**
     * Edit Component includes multiple sliders
     */
    public Edit = {
        view: (vnode) => {
 
            console.log(vnode.attrs)
            let ftAvail = vnode.attrs.requested - vnode.attrs.parttime;
            let ptAvail = vnode.attrs.requested - vnode.attrs.fulltime;
            return m('div.border.border-blue.bg-white.rounded.p2.mx2.flex.flex-column', 
            [
                UIProgressBar(ftAvail, vnode.attrs.requested, "Big Widget slots: " + ftAvail),
                UIProgressBar(vnode.attrs.fulltime, vnode.attrs.requested),
                UISlider('Big Widget', vnode.attrs.fulltime, (val) => this.app.actions.changeAction('fulltime', val), vnode.attrs.requested),
                
                UIProgressBar(ptAvail, vnode.attrs.requested, "Small Widget slots: " + ptAvail ),
                UIProgressBar(vnode.attrs.parttime, vnode.attrs.requested),
                UISlider('Small Widget', vnode.attrs.parttime, (val) => this.app.actions.changeAction('parttime', val), vnode.attrs.requested),

                m(this.Footer, vnode.attrs)
            ]);
        }
    } as m.Component<AssignmentProps, VState>
    
    
    /**
     * Footer shows schedule versus total and a checkmark icon
     */
    public Footer = {
        view: (vnode) => {
            
            let matchClass = vnode.attrs.isValid ? 'icon-ok' : '';
            return m('div.p2.m1',
                {
                    class: vnode.attrs.isValid ? 'bg-green' : 'bg-red'
                },
                'Widget Quota: ' + vnode.attrs.total + ' of ' + vnode.attrs.requested, m('span', { class: matchClass }))
        }
    } as m.Component<AssignmentProps, VState>

    /**
     * Review Component 
     */
    public Review = {
        view: (vnode) => {
           
            return m("div.border.mx2", [
                m("table.table", [
                    m("tbody", [
                        m("tr", [
                            m("td", "Big Widget Slots"),
                            m("td", vnode.attrs.fulltime)
                        ]),
                        m("tr", [
                            m("td", "Small Widget Slots"),
                            m("td", vnode.attrs.parttime)
                        ]),
                        m("tr.text-bold.bg-blue", [
                            m("td", "Total"),
                            m("td", [
                                m(this.Footer, vnode.attrs)
                            ])
                        ])
                    ])
                ]),
                m('button.btn.btn-lg.m3',
                {
                    class: vnode.attrs.isValid ? 'bg-blue': 'bg-gray',
                    disabled: !vnode.attrs.isValid ? 'disabled' : '',
                    onclick: (e) => this.app.actions.saveData()
                },
                'Save')
            ])
        }
        
    } as m.Component<AssignmentProps, VState>

    
}