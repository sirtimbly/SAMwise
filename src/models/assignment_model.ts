import {Model} from '../samwise';
/** 
 * AssignmentProps
 */
export class AssignmentProps {
    fulltime: number;
    parttime: number;
    requested: number;
    total: number;
    isEditing: true;
    isValid: boolean;

    constructor (data?: any) {
        if (data) {
            this.fulltime = data.fulltime || 0;
            this.parttime = data.parttime || 0;
            this.requested = data.requested || 0;
            this.total = data.total || 0;
            this.isEditing = data.isEditing || 0;
        }
    }
}

/**
 * AssignmentModel
 */
export class AssignmentModel implements Model {
    
    props: AssignmentProps;
    // the renderer method will be implementation specific each time this model is used
    stateRenderer: (data:AssignmentProps) => void; 

    constructor () {
        this.props = new AssignmentProps({
            fulltime:20,
            parttime:30,
            requested:80
        });
    }

    present = (data: any) => {  
        let data_ = new AssignmentProps();
        if (data !== undefined) {
            data_.fulltime = data.fulltime || this.props.fulltime;
            data_.parttime = data.parttime || this.props.parttime;
            data_.requested = data.requested || this.props.requested;
            if (data.isEditing !== undefined) {
                this.props.isEditing = data.isEditing;
            }
        }
        var newTotal = Number(data_.fulltime) + Number(data_.parttime);
        console.log("newtotal", newTotal)
        if (newTotal <= this.props.requested) {
            this.props.fulltime = data_.fulltime;
            this.props.parttime = data_.parttime;
        } else {
            if (data.fulltime && !data.parttime) {
                this.props.fulltime = data_.fulltime - (newTotal - this.props.requested);
            }
            if (data.parttime && !data.fulltime) {
                this.props.parttime = data_.parttime - (newTotal - this.props.requested);
            }
            
        }

        //always recalculate total
        this.props.total = this.props.fulltime + this.props.parttime;
        //finally we close the functional reactive loop
        this.stateRenderer(this.props)
    }

    
}
