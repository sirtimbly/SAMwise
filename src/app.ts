import { AssignmentModel, AssignmentProps } from './models/assignment_model';
import { App, State, Model, Action } from './samwise';
import * as m from 'mithril';
import { AppViewComponents } from './components/AppViewComponents'

// let m = Mithril;
let app = new App;

app.model = new AssignmentModel();

export class AppViewActions {
    /**
     * Actions (the UI sends messages out to change the model)
     * - changeAction() updates the model mased on slider events, the present() method has a default value so we don't have to leak model coupling into our views
     * - loadAction() gets new data will usually just be called when the app is first initialized
     */
    changeAction = function(target:string, value:number, present: (data: AssignmentProps) => void = app.model.present) : void {
            var data_ = new AssignmentProps();
            data_[target] = Number(value);
            present(data_);
    }
    loadAction = function(present: (data:AssignmentProps) => void) {
            console.log("load: pretend we have an api call here");
            present(new AssignmentProps({
            fulltime:20,
            parttime:30,
            requested:80,
            isEditing: true
            }));
    }
    switchToReview = function(present: (data:AssignmentProps) => void = app.model.present) : void {
        console.log('switching tabs')
        present(new AssignmentProps({isEditing: false}))
    }
    switchToEdit = function(present: (data:AssignmentProps) => void = app.model.present) : void {
        console.log('switching tabs')
        present(new AssignmentProps({isEditing: true}))
    }

    saveData = function(present: (data:AssignmentProps) => void = app.model.present) : void {
        console.log('save clicked, fake saving to an API');
    }
}
app.actions = new AppViewActions();


/* ============================================ */

/**
 * AppState (when the model updates, examine it and determine how to update the UI)
 * - editing
 * - reviewing
 */
class AppState implements State {
    render = (model: AssignmentProps) => {
        
        if (model.total > model.requested) {
            console.log("do some error checking and warnings here - like a toast");
        }

        model.isValid = (model.total === model.requested);
        this.representation(model);
        this.nextAction(model);
        
    }
    representation = (model: AssignmentProps) => {
        m.render(document.getElementById('tabs'), m(app.views.Tabs, model));
        
        //depending on which state the app is in we show a different component
        let mainScreenComponent = (model.isEditing) ? app.views.Edit : app.views.Review;

        m.render(document.getElementById('mainapp'), m(mainScreenComponent, model));
    }
    nextAction = (model: AssignmentProps) => {
       // throw new Error('Method not implemented.');
    }
}

app.state = new AppState();

//setting the default model stateRenderer to state function is important for decoupling state implementation from the model implementation
app.model.stateRenderer = app.state.render;

app.views = new AppViewComponents(app);
export const Application = app;
/* ============================================ */
/* ENGAGE! */

console.log("initial view state");
app.actions.loadAction(app.model.present)