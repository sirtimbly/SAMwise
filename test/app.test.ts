import {tidy} from '../test-helper'
import { AppViewComponents} from '../src/components/AppViewComponents'
import { AssignmentModel, AssignmentProps } from '../src/models/assignment_model';
import { App, State, Model, Action } from '../src/samwise';
import * as m from 'mithril';
import mq from 'mithril-query'


describe("App", () => {
    it("should have footer", () => {
        var app = new App
        var components = new AppViewComponents(app);
        var cmp = m(components.Edit, <AssignmentProps>{
            fulltime:20,
            parttime:30,
            requested:80,
            isEditing: true
            });
        expect(cmp)
        //var output = mq(cmp);
        //expect(output.contains('50 of 80'));
        const html = tidy(cmp);
        expect(html).toMatchSnapshot();
    });
})