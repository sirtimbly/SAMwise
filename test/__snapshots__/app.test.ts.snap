// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should have footer 1`] = `
"<div class=
\\"border border-blue bg-white rounded p2 mx2 flex flex-column\\">
  <div class=\\"guage-item flex px2\\">
    <progress value=\\"50\\" max=\\"80\\" class=
    \\"mx1 col-8\\"></progress><span class=\\"bold ml2\\">Big Widget slots:
    50</span>
  </div>
  <div class=\\"guage-item flex px2\\">
    <progress value=\\"20\\" max=\\"80\\" class=\\"mx1 col-8\\"></progress>
  </div>
  <div class=\\"flex bg-light-gray p2 m1 flex-reverse\\">
    <input type=\\"range\\" min=\\"0\\" max=\\"80\\" name=\\"range_Big%20Widget\\"
    class=\\"col-8\\"><label for=\\"range_Big%20Widget\\" class=
    \\"title col-4 px2\\">20 Big Widget shifts assigned.</label>
  </div>
  <div class=\\"guage-item flex px2\\">
    <progress value=\\"60\\" max=\\"80\\" class=
    \\"mx1 col-8\\"></progress><span class=\\"bold ml2\\">Small Widget
    slots: 60</span>
  </div>
  <div class=\\"guage-item flex px2\\">
    <progress value=\\"30\\" max=\\"80\\" class=\\"mx1 col-8\\"></progress>
  </div>
  <div class=\\"flex bg-light-gray p2 m1 flex-reverse\\">
    <input type=\\"range\\" min=\\"0\\" max=\\"80\\" name=
    \\"range_Small%20Widget\\" class=\\"col-8\\"><label for=
    \\"range_Small%20Widget\\" class=\\"title col-4 px2\\">30 Small Widget
    shifts assigned.</label>
  </div>
  <div class=\\"p2 m1 bg-red\\">
    Widget Quota: undefined of 80<span class=\\"\\"></span>
  </div>
</div>
"
`;
